### PR TITLE
[PERF] test_website_slides_full: certification tour is slow

### PR DESCRIPTION
**Pull request by yagp-odoo**
Currently 'certification_member' tour is too slow that We had to increase the
default timeout of 60 seconds.

The purpose of this PR is to reduce tour time by removing steps / decrease fail
attempt or optimizing tour steps.

Changes:

As the purpose of the test/tour is to check course certification flow,
**NOT** to test the ecommerce checkout flow, which is taking too much time.

We removed course purchase flow and used python to create sale order for the member.

--> Python prepares certification data (course / survey)
--> Python creates a sale order for the course and confirms it, should add the
member to the course (so we do the steps in python rather than in the tour
inside the ecommerce shop)
--> Python runs test first,
--> The tour fails the certification 3 times,
--> Python checks that the member is kicked out of the course.
--> Python creates a second sale order and confirms it, should add the member again
and run it,
--> The tour succeeds the certification (with the profile steps at the end)
--> Python do some extra bonus checks, verifying membership status, course completion..

Improvement:

Tour average runtime (Before this PR): **~59-62 seconds**
Tour average runtime (After this PR):  **~36-39 seconds**
Performance Improvement: **~38%**

Task-3478117

### CODE REVIEW

File: addons/test_website_slides_full/static/tests/tours/slides_certification_member.js [hunk 1]
```diff
@@ -50,16 +49,22 @@ var buyCertificationSteps = [{
...
+    run: 'click',
+}, {
+    content: 'eCommerce: add card number',
+    trigger: 'input[name="customer_input"]',
+    run: "edit 4343434343434343",
+},
+    tourUtils.pay(),
+{
+    content: 'eCommerce: check that the payment is successful',
+    trigger: '.oe_website_sale_tx_status:contains("Your payment has been successfully processed.")',
```
Comment (thsh-odoo) [inline addons/test_website_slides_full/static/tests/tours/slides_certification_member.js:64]: Why do we need to write this code again?
we have the function?

File: addons/test_website_slides_full/static/tests/tours/slides_certification_member.js [hunk 2]
```diff
@@ -50,16 +49,22 @@ var buyCertificationSteps = [{
     run: "click",
 },
     tourUtils.goToCart(),
-    tourUtils.goToCheckout(),
-    ...tourUtils.payWithDemo(),
-    wTourUtils.clickOnExtraMenuItem({}),
```
Comment (thsh-odoo) [inline addons/test_website_slides_full/static/tests/tours/slides_certification_member.js:55]: Why did you remove this any reasons?

**Review by thsh-odoo — COMMENTED** [review]
Hello @yagp-odoo 
Left few questions
Thanks!! :smiley:

File: addons/test_website_slides_full/static/tests/tours/slides_certification_member.js (hunk 1 above)
Comment (yagp-odoo) [reply addons/test_website_slides_full/static/tests/tours/slides_certification_member.js:64]: `payWithDemo` function only works on `/shop/payment` page, and we're not going there as we have option to pay directly from `/shop/cart`.

**Comment by awa-odoo** [comment]
Hello @yagp-odoo 

One more idea to try to reduce the timing.

Inside this tour: https://github.com/odoo/odoo/blob/master/addons/survey/static/tests/tours/certification_failure.js#L10
We patch the survey to remove fade in / out delays. Maybe that can help?
Although I see that the survey is configured as "one page", so not sure if it's going to change that much.

Did you identify where we are loosing the most time? Can't it be further improved (without changing too much code)?

**Comment by yagp-odoo** [comment]
Hi, @awa-odoo :wave:

> We patch the survey to remove fade in / out delays. Maybe that can help?
Although I see that the survey is configured as "one page", so not sure if it's going to change that much.

In survey we added [extra fadeInOut effect](https://github.dev/odoo/odoo/blob/c08db2dbdcb4cb386e3ad27b249425b243dd0b12/addons/survey/static/src/js/survey_form.js#L47-L48), and that survey tour has too many question compare to our slide tour. I don't think it will affect that much..(If you see tour with `watch=True` the question page will quickly filled up, that doesn't take too much time in first place.)

> Did you identify where we are loosing the most time? Can't it be further improved (without changing too much code)?

TBH, purchasing course takes most of the time, I did reduce it quite a bit tho, If we don't reduce the attempt it will takes around 52-55 Sec. (It's taking ~48-52 Sec to complete with 2 attempts :(, it changes every time I run the tour _~sigh_)

So, if we don't want to change fail attempts, we can just improve code for purchasing course...

**Comment by awa-odoo** [comment]
Hello @yagp-odoo 

If the part purchasing the course takes the most time, then there is one ore thing we can try, but it's going to require a bit of extra work.

What we do sometimes is split the test into multiple consequent tours.
So you have:
- Some bits of python code preparing data / doing some checks
- A first tour running
- Some python checks
- A second tour running
- Some more python checks
- ...

A good example is this one:
https://github.com/odoo/odoo/blob/master/addons/survey/tests/test_survey_ui_session.py#L14

As the goal of the tour you are working on is NOT to test the ecommerce checkout flow, we could break it down into the following:
- Python prepares certification data (course / survey, as currently)
- Python creates a sale order for the course and confirms it, should add the member to the course (so we do the steps in python rather than in the tour inside the ecommerce shop)
- Python runs a first tour
- The tour fails the certification twice
- Python checks that the member is kicked out of the course
- Python creates a second sale order and confirms it, should add the member again
- Python runs a second tour
- The tour succeeds the certification (with the profile steps at the end)
- Python can do some extra bonus checks, verifying membership status, course completion etc

As said, this is a significant rework of this tour / test, but I feel like it could be worth it to save run time and would even make the test suite more robust by adding various python checks.

Then we can time the python testing method and make sure it takes a lot less time than currently.

What do you think? Are you willing to give it a try?

Let me know if this is unclear.

Thanks & cheers!

**Comment by yagp-odoo** [comment]
Hello, @awa-odoo :wave:

do we still needs this ` # Avoid Billing/Shipping address page` & `# Specify Accounting Data` ? I checked and it's working without them...

Thanks!

File: addons/test_website_slides_full/tests/test_ui_wslides.py [hunk 3]
```diff
@@ -168,4 +146,58 @@ def test_course_certification_employee(self):
...
+        self.env['sale.order'].create({
+            'partner_id': user_demo.partner_id.id,
+            'order_line': [
+                (0, 0, {
+                    'name': furniture_course_product.name,
+                    'product_id': furniture_course_product.id,
+                    'product_uom_qty': 1,
+                    'price_unit': furniture_course_product.list_price,
+                })
+            ],
```
Comment (awa-odoo) [inline addons/test_website_slides_full/tests/test_ui_wslides.py:185]: Could put those in a variable as used twice.

File: addons/test_website_slides_full/tests/test_ui_wslides.py [hunk 4]
```diff
@@ -12,56 +12,21 @@
 class TestUi(AccountTestInvoicingCommon, TestUICommon):
```
Comment (awa-odoo) [inline addons/test_website_slides_full/tests/test_ui_wslides.py:12]: Do we still need "AccountTestInvoicingCommon" ?

To answer your question on the PR, I would indeed clean and remove everything we don't need :slightly_smiling_face:

**Review by awa-odoo — COMMENTED** [review]
Hello @yagp-odoo 

Thanks for the rework, looking good!
I left a few comments but this is going in the right direction.

How are we looking on running time now by the way? Did it help significantly?

File: addons/test_website_slides_full/tests/test_ui_wslides.py [hunk 5]
```diff
@@ -168,4 +146,58 @@ def test_course_certification_employee(self):
...
+                (0, 0, {
+                    'name': furniture_course_product.name,
+                    'product_id': furniture_course_product.id,
+                    'product_uom_qty': 1,
+                    'price_unit': furniture_course_product.list_price,
+                })
+            ],
+        }).action_confirm()
+        # Check Member has access to the course
+        self.assertIn(user_demo.partner_id, furniture_course.partner_ids)
```
[the remark this answers is not in the document]
Comment (yagp-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py]: I think "Demo" has the rights to create survey/course/product as when I tested it was not giving me any access error.

IMO, we can use "portal user"? as this test/tour is to check flow for public users(i.e., student here)?

> How are we looking on running time now by the way? Did it help significantly?

It is!!  I say the runtime is reduce to aprox half without accounting data...

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py]: > IMO, we can use "portal user"? as this test/tour is to check flow for public users(i.e., student here)?

Portal would be good if possible yes, but I would not overcomplicate the code to make it work.
If it's much more simple to use "demo", we can go that route 🙂

[the remark this answers is not in the document]
Comment (yagp-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py]: Well 'Demo' has most of rights we need for purchasing course, so we don't need any `sudo` it should be simple, but i just thought if we need a more 'practical use case' we can use 'portal' :smiley:

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py]: As said, if it's possible while keeping the code simple, go for it, it will indeed be a more common case :slightly_smiling_face:

**Comment by yagp-odoo** [comment]
Hello, @awa-odoo :wave: 
I have updated the code, you can give remarks whenever you have some time..
Thank you :smiley:

File: addons/test_website_slides_full/tests/test_ui_wslides.py [hunk 6]
```diff
@@ -163,9 +131,53 @@ def test_course_certification_employee(self):
...
+                    'price_unit': self.furniture_course_product.list_price,
+                })
+            ]}
+
+        # =================
+        # FAILURE ATTEMPTS
+        # =================
+        self.env['sale.order'].sudo().create(sale_order_data).action_confirm()
+        # Member should have access to the course
+        self.assertTrue(self.furniture_course.with_user(user_portal).is_member)
```
Comment (awa-odoo) [inline addons/test_website_slides_full/tests/test_ui_wslides.py:159]: Should not have to use "with_user"as the test annotation member is already portal.

(Same for others here below).

If it doesn't work otherwise, maybe try invalidating the cache of "self.furniture_course" at the beginning of the method.

File: addons/test_website_slides_full/tests/test_ui_wslides.py [hunk 7]
```diff
@@ -4,66 +4,21 @@
...
-            'reconcile': True,
-            'account_type': 'asset_receivable',
-        })
-        a_pay = self.env['account.account'].create({
-            'code': 'X1111',
-            'name': 'Creditors - (test)',
-            'account_type': 'liability_payable',
-            'reconcile': True,
-        })
+    def setUp(cls):
```
Comment (awa-odoo) [inline addons/test_website_slides_full/tests/test_ui_wslides.py:14]: We usually always use "setUpClass", even if in this case it should not change anything, mostly best practice.

File: addons/test_website_slides_full/__manifest__.py [hunk 8]
```diff
@@ -13,7 +13,6 @@
         'website_sale_slides',
```
Comment (awa-odoo) [inline addons/test_website_slides_full/__manifest__.py:13]: Commit message:

As the purpose of this commit is performance, it's important to state how long it's taking after the changes (on average).
(You can also specify the percentage of improvement, always nice to see).

**Review by awa-odoo — COMMENTED** [review]
Hello @yagp-odoo 

Almost there on this one, only a few minor polishing comments left.

Thanks for your patience & cheers!

File: addons/test_website_slides_full/tests/test_ui_wslides.py (hunk 7 above)
Comment (yagp-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py:14]: I understand but, we're using this `TestUICommon` and they used `setUp` , See [Here](https://github.dev/odoo/odoo/blob/a56f395e9c012d9b4afa342ab2c4826b23083461/addons/website_slides/tests/test_ui_wslides.py#L15-L18)

Comment (awa-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py:14]: Fair enough, you can keep it then.

File: addons/test_website_slides_full/tests/test_ui_wslides.py (hunk 6 above)
Comment (yagp-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py:159]: It's not working (even with invalidate cache), it was always giving `False`...
image

Comment (yagp-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py:159]: For test purpose, I checked env user in test function, and I was getting `OdooBot` for some reason....

```py
    @users("portal")
    def test_course_certification_employee(self):
        user_portal = self.env.user
        # self.furniture_course._invalidate_cache()
        test = self.furniture_course.action_redirect_to_certified_members()
        # breakpoint is in action
```
image

I'm not sure why it's happening.. :confused:

Comment (awa-odoo) [reply addons/test_website_slides_full/tests/test_ui_wslides.py:159]: What if you try at the beginning of the function:

`self.furniture_course = self.env['slide.channel'].browse(self.furniture_course.id)  # use proper environment to test user dependent computes`

If it doesn't work, let's just keep the with_user then, it's fine.

(Don't forget to mention the new timings in the commit message so we can merge this one).

Thanks & cheers!

File: addons/test_website_slides_full/__manifest__.py (hunk 8 above)
Comment (yagp-odoo) [reply addons/test_website_slides_full/__manifest__.py:13]: @awa-odoo Done. :smile:

**Comment by awa-odoo** [comment]
@robodoo r+
